src/portalnet/protocol.rs

```diff
@@ -32,11 +33,74 @@ impl Default for PortalnetConfig {
 pub const PROTOCOL: &str = "portal";
 
 pub struct PortalnetProtocol {
-    discovery: Discovery,
+    discovery: Arc<Discovery>,
     data_radius: U256,
+}
+
+pub struct PortalnetEvents {
+    data_radius: U256,
+    discovery: Arc<Discovery>,
     protocol_receiver: mpsc::UnboundedReceiver<TalkRequest>,
 }
 
+impl PortalnetEvents {
+    /// Receives a request from the talkreq handler and sends a response back
+    pub async fn process_requests(mut self) {
+        while let Some(request) = self.protocol_receiver.recv().await {
+            debug!("Got talkreq message {:?}", request);
+            let reply = match self.process_one_request(&request).await {
+                Ok(r) => Message::Response(r).to_bytes(),
+                Err(e) => {
+                    error!("failed to process portal event: {}", e);
+                    e.into_bytes()
+                }
+            };
+
+            request.respond(reply);
+        }
+    }
+
+    async fn process_one_request(&self, talk_request: &TalkRequest) -> Result<Response, String> {
+        let protocol = std::str::from_utf8(talk_request.protocol())
+            .map_err(|_| "Invalid protocol".to_owned())?;
+
+        if protocol != PROTOCOL {
+            return Err("Invalid protocol".to_owned());
+        }
+
+        let request = match Message::from_bytes(talk_request.body()) {
+            Ok(Message::Request(r)) => r,
+            Ok(_) => return Err("Invalid message".to_owned()),
+            Err(e) => return Err(format!("Invalid request: {}", e)),
+        };
+
+        let response = match request {
+            Request::Ping(Ping { .. }) => {
+                debug!("Got overlay ping request {:?}", request);
+                let enr_seq = self.discovery.local_enr().seq();
+                Response::Pong(Pong {
+                    enr_seq: enr_seq,
+                    data_radius: self.data_radius,
+                })
+            }
+
+            Request::FindNodes(FindNodes { distances }) => {
+                let enrs = self.discovery.find_nodes_response(distances);
```

Comment:
I think we can't simply reuse the discv5 find_nodes machinery here, because that will be populated by *all* nodes on the network (not necessarily only portal nodes). We only want to return nodes that are definitely in the portal network. I think we'll have to have a whole other copy of the buckets that are populated only by the portal nodes.

That seems like a big enough thing to go into its own PR, though.

Reply:
Yeah, I think that's the right approach. I was just copying what was there before my PR.